File: src/lib/insights-api.ts
import type { Account, Budget, Category, Goal, RecurringTransaction, Transaction } from '@/types'

export type InsightType =
  | 'spending'
  | 'trend'
  | 'anomaly'
  | 'budget'
  | 'behavior'
  | 'savings'
  | 'recurring'
  | 'recommendation'

export type InsightSeverity = 'info' | 'positive' | 'warning' | 'critical'

export interface FinancialInsight {
  id: string
  type: InsightType
  title: string
  summary: string
  explanation: string
  severity: InsightSeverity
  confidence: number
  impact_score: number
  metrics: Record<string, number | string>
  category: string | null
  related_transaction_ids: string[]
  recommendation: string | null
  source?: 'deterministic' | 'llm'
}

export interface AnalyzeResponse {
  summary: string
  financial_health_summary: string
  insights: FinancialInsight[]
  metrics: Record<string, unknown>
  generated_at: string
  analysis_period: {
    start: string
    end: string
    comparison_start: string
    comparison_end: string
    label: string
  }
  used_fallback: boolean
  llm_available: boolean
}

export interface RelatedTransaction {
  id: string
  merchant: string
  amount: number
  date: string
  category: string | null
}

export interface ChatMetric {
  id: string
  label: string
  value: number
  previous: number | null
  change: number | null
  unit: 'money' | 'percent' | string
}

export interface VioChartSeries {
  type: 'line' | 'bar'
  points: Array<{ label: string; value: number }>
}

export interface VioChatMessage {
  id: string
  role: 'user' | 'assistant'
  content: string
  insights: FinancialInsight[]
  metrics: ChatMetric[]
  related_transactions: RelatedTransaction[]
  chart: VioChartSeries | null
  follow_ups: string[]
  grounding: string | null
  created_at: string
}

export interface ConversationSummary {
  id: string
  title: string
  updated_at: string
  message_count: number
}

export interface ConversationRecord {
  id: string
  title: string
  messages: VioChatMessage[]
  updated_at: string
}

export interface ChatResponse {
  conversation_id: string
  answer: string
  message: VioChatMessage | null
  insights: FinancialInsight[]
  metrics: Record<string, unknown>
  used_fallback: boolean
  empty_data: boolean
}

export interface FinanceSnapshotPayload {
  currency: string
  transactions: Transaction[]
  categories: Category[]
  accounts: Account[]
  budgets: Budget[]
  goals: Goal[]
  recurring: RecurringTransaction[]
}

const baseUrl = (import.meta.env.VITE_INSIGHTS_API_URL ?? '/api').replace(/\/$/, '')

function dataVersion(snapshot: FinanceSnapshotPayload): string {
  const stamps = [
    ...snapshot.transactions,
    ...snapshot.budgets,
    ...snapshot.accounts,
    ...snapshot.goals,
  ].map((item) => item.updatedAt)
  return `${stamps.sort().at(-1) ?? 'empty'}:${snapshot.transactions.length}`
}

export function insightsHeaders(accessToken: string, userId: string): HeadersInit {
  return {
    Authorization: `Bearer ${accessToken}`,
    'Content-Type': 'application/json',
    'X-User-Id': userId,
  }
}

export async function analyzeInsights(input: {
  accessToken: string
  userId: string
  startDate: string
  endDate: string
  snapshot: FinanceSnapshotPayload
}): Promise<AnalyzeResponse> {
  const response = await fetch(`${baseUrl}/insights/analyze`, {
    method: 'POST',
    headers: insightsHeaders(input.accessToken, input.userId),
    body: JSON.stringify({
      start_date: input.startDate,
      end_date: input.endDate,
      snapshot: input.snapshot,
      data_version: dataVersion(input.snapshot),
      currency: input.snapshot.currency,
    }),
  })
  if (!response.ok) throw new Error(await errorMessage(response))
  return response.json() as Promise<AnalyzeResponse>
}

export async function streamChatInsights(
  input: {
    accessToken: string
    userId: string
    message: string
    conversationId?: string
    startDate: string
    endDate: string
    snapshot: FinanceSnapshotPayload
    page?: string
  },
  onProgress?: (message: string) => void,
): Promise<ChatResponse> {
  const body = {
    message: input.message,
    conversation_id: input.conversationId,
    start_date: input.startDate,
    end_date: input.endDate,
    snapshot: input.snapshot,
    page: input.page,
  }
  try {
    const response = await fetch(`${baseUrl}/insights/chat/stream`, {
      method: 'POST',
      headers: insightsHeaders(input.accessToken, input.userId),
      body: JSON.stringify(body),
    })
    if (!response.ok || !response.body) throw new Error(await errorMessage(response))
    const reader = response.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ''
    let result: ChatResponse | null = null
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      const chunks = buffer.split('\n\n')
      buffer = chunks.pop() ?? ''
      for (const chunk of chunks) {
        const event = chunk.match(/^event:\s*(\w+)/m)?.[1]
        const dataLine = chunk.split('\n').find((line) => line.startsWith('data:'))
        if (!dataLine) continue
        const data = JSON.parse(dataLine.replace(/^data:\s*/, '')) as ChatResponse & { message?: string }
        if (event === 'progress' && data.message) onProgress?.(String(data.message))
        if (event === 'complete') result = data
      }
    }
    if (result) return result
  } catch {
    /* Fall through to the non-streaming chat endpoint. */
  }
  return chatInsights(input)
}

export async function chatInsights(input: {
  accessToken: string
  userId: string
  message: string
  conversationId?: string
  startDate: string
  endDate: string
  snapshot: FinanceSnapshotPayload
  page?: string
}): Promise<ChatResponse> {
  const response = await fetch(`${baseUrl}/insights/chat`, {
    method: 'POST',
    headers: insightsHeaders(input.accessToken, input.userId),
    body: JSON.stringify({
      message: input.message,
      conversation_id: input.conversationId,
      start_date: input.startDate,
      end_date: input.endDate,
      snapshot: input.snapshot,
      page: input.page,
    }),
  })
  if (!response.ok) throw new Error(await errorMessage(response))
  return response.json() as Promise<ChatResponse>
}

export async function getLatestInsights(accessToken: string, userId: string): Promise<AnalyzeResponse | null> {
  const response = await fetch(`${baseUrl}/insights/latest`, {
    headers: insightsHeaders(accessToken, userId),
  })
  if (!response.ok) return null
  return response.json() as Promise<AnalyzeResponse | null>
}

export async function listVioConversations(accessToken: string, userId: string): Promise<ConversationSummary[]> {
  const response = await fetch(`${baseUrl}/insights/conversations`, {
    headers: insightsHeaders(accessToken, userId),
  })
  if (!response.ok) throw new Error(await errorMessage(response))
  return response.json() as Promise<ConversationSummary[]>
}

export async function getVioConversation(
  accessToken: string,
  userId: string,
  conversationId: string,
): Promise<ConversationRecord> {
  const response = await fetch(`${baseUrl}/insights/conversations/${conversationId}`, {
    headers: insightsHeaders(accessToken, userId),
  })
  if (!response.ok) throw new Error(await errorMessage(response))
  return response.json() as Promise<ConversationRecord>
}

export async function createVioConversation(accessToken: string, userId: string): Promise<ConversationRecord> {
  const response = await fetch(`${baseUrl}/insights/conversations`, {
    method: 'POST',
    headers: insightsHeaders(accessToken, userId),
  })
  if (!response.ok) throw new Error(await errorMessage(response))
  return response.json() as Promise<ConversationRecord>
}

export async function deleteVioConversation(accessToken: string, userId: string, conversationId: string): Promise<void> {
  const response = await fetch(`${baseUrl}/insights/conversations/${conversationId}`, {
    method: 'DELETE',
    headers: insightsHeaders(accessToken, userId),
  })
  if (!response.ok) throw new Error(await errorMessage(response))
}

export async function sendInsightFeedback(input: {
  accessToken: string
  userId: string
  insightId: string
  feedback: 'helpful' | 'not_helpful' | 'not_relevant' | 'already_know'
}): Promise<void> {
  const response = await fetch(`${baseUrl}/insights/${input.insightId}/feedback`, {
    method: 'POST',
    headers: insightsHeaders(input.accessToken, input.userId),
    body: JSON.stringify({ feedback: input.feedback }),
  })
  if (!response.ok) throw new Error(await errorMessage(response))
}

export async function invalidateInsightsCache(accessToken: string, userId: string): Promise<void> {
  await fetch(`${baseUrl}/insights/cache/invalidate`, {
    method: 'POST',
    headers: insightsHeaders(accessToken, userId),
  })
}

async function errorMessage(response: Response): Promise<string> {
  try {
    const body = (await response.json()) as { detail?: string }
    return body.detail ?? `Insights request failed (${response.status})`
  } catch {
    return `Insights request failed (${response.status})`
  }
}

File: flutter/lib/data/insights_api.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/env.dart';
import '../models/models.dart';

class FinancialInsight {
  const FinancialInsight({
    required this.id,
    required this.type,
    required this.title,
    required this.summary,
    required this.explanation,
    required this.severity,
    required this.confidence,
    required this.impactScore,
    required this.metrics,
    required this.category,
    required this.relatedTransactionIds,
    required this.recommendation,
    required this.source,
  });

  final String id;
  final String type;
  final String title;
  final String summary;
  final String explanation;
  final String severity;
  final double confidence;
  final double impactScore;
  final Map<String, dynamic> metrics;
  final String? category;
  final List<String> relatedTransactionIds;
  final String? recommendation;
  final String source;

  factory FinancialInsight.fromJson(Map<String, dynamic> json) {
    return FinancialInsight(
      id: json['id'] as String? ?? '',
      type: json['type'] as String? ?? 'spending',
      title: json['title'] as String? ?? '',
      summary: json['summary'] as String? ?? '',
      explanation: json['explanation'] as String? ?? '',
      severity: json['severity'] as String? ?? 'info',
      confidence: (json['confidence'] as num?)?.toDouble() ?? 0.8,
      impactScore: (json['impact_score'] as num?)?.toDouble() ?? 0,
      metrics: Map<String, dynamic>.from(json['metrics'] as Map? ?? {}),
      category: json['category'] as String?,
      relatedTransactionIds:
          (json['related_transaction_ids'] as List?)?.map((item) => '$item').toList() ?? const [],
      recommendation: json['recommendation'] as String?,
      source: json['source'] as String? ?? 'deterministic',
    );
  }
}

class AnalyzeResponse {
  const AnalyzeResponse({
    required this.summary,
    required this.financialHealthSummary,
    required this.insights,
    required this.metrics,
    required this.generatedAt,
    required this.usedFallback,
    required this.llmAvailable,
  });

  final String summary;
  final String financialHealthSummary;
  final List<FinancialInsight> insights;
  final Map<String, dynamic> metrics;
  final String generatedAt;
  final bool usedFallback;
  final bool llmAvailable;

  factory AnalyzeResponse.fromJson(Map<String, dynamic> json) {
    return AnalyzeResponse(
      summary: json['summary'] as String? ?? '',
      financialHealthSummary: json['financial_health_summary'] as String? ?? '',
      insights: (json['insights'] as List? ?? [])
          .map((item) => FinancialInsight.fromJson(Map<String, dynamic>.from(item as Map)))
          .toList(),
      metrics: Map<String, dynamic>.from(json['metrics'] as Map? ?? {}),
      generatedAt: json['generated_at']?.toString() ?? '',
      usedFallback: json['used_fallback'] as bool? ?? false,
      llmAvailable: json['llm_available'] as bool? ?? false,
    );
  }
}

class ChatMetric {
  const ChatMetric({
    required this.id,
    required this.label,
    required this.value,
    required this.previous,
    required this.change,
    required this.unit,
  });

  final String id;
  final String label;
  final double value;
  final double? previous;
  final double? change;
  final String unit;

  factory ChatMetric.fromJson(Map<String, dynamic> json) {
    return ChatMetric(
      id: json['id'] as String? ?? '',
      label: json['label'] as String? ?? '',
      value: (json['value'] as num?)?.toDouble() ?? 0,
      previous: (json['previous'] as num?)?.toDouble(),
      change: (json['change'] as num?)?.toDouble(),
      unit: json['unit'] as String? ?? 'money',
    );
  }
}

class RelatedTransaction {
  const RelatedTransaction({
    required this.id,
    required this.merchant,
    required this.amount,
    required this.date,
    required this.category,
  });

  final String id;
  final String merchant;
  final double amount;
  final String date;
  final String? category;

  factory RelatedTransaction.fromJson(Map<String, dynamic> json) {
    return RelatedTransaction(
      id: json['id'] as String? ?? '',
      merchant: json['merchant'] as String? ?? '',
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      date: json['date'] as String? ?? '',
      category: json['category'] as String?,
    );
  }
}

class VioChatMessage {
  const VioChatMessage({
    required this.id,
    required this.role,
    required this.content,
    required this.insights,
    required this.metrics,
    required this.relatedTransactions,
    required this.followUps,
    required this.grounding,
    required this.createdAt,
  });

  final String id;
  final String role;
  final String content;
  final List<FinancialInsight> insights;
  final List<ChatMetric> metrics;
  final List<RelatedTransaction> relatedTransactions;
  final List<String> followUps;
  final String? grounding;
  final String createdAt;

  factory VioChatMessage.fromJson(Map<String, dynamic> json) {
    return VioChatMessage(
      id: json['id'] as String? ?? '',
      role: json['role'] as String? ?? 'assistant',
      content: json['content'] as String? ?? '',
      insights: (json['insights'] as List? ?? [])
          .map((item) => FinancialInsight.fromJson(Map<String, dynamic>.from(item as Map)))
          .toList(),
      metrics: (json['metrics'] as List? ?? [])
          .map((item) => ChatMetric.fromJson(Map<String, dynamic>.from(item as Map)))
          .toList(),
      relatedTransactions: (json['related_transactions'] as List? ?? [])
          .map((item) => RelatedTransaction.fromJson(Map<String, dynamic>.from(item as Map)))
          .toList(),
      followUps: (json['follow_ups'] as List?)?.map((item) => '$item').toList() ?? const [],
      grounding: json['grounding'] as String?,
      createdAt: json['created_at']?.toString() ?? '',
    );
  }
}

class ChatResponse {
  const ChatResponse({
    required this.conversationId,
    required this.answer,
    required this.message,
    required this.insights,
    required this.usedFallback,
    required this.emptyData,
  });

  final String conversationId;
  final String answer;
  final VioChatMessage? message;
  final List<FinancialInsight> insights;
  final bool usedFallback;
  final bool emptyData;

  factory ChatResponse.fromJson(Map<String, dynamic> json) {
    return ChatResponse(
      conversationId: json['conversation_id'] as String? ?? '',
      answer: json['answer'] as String? ?? '',
      message: json['message'] is Map
          ? VioChatMessage.fromJson(Map<String, dynamic>.from(json['message'] as Map))
          : null,
      insights: (json['insights'] as List? ?? [])
          .map((item) => FinancialInsight.fromJson(Map<String, dynamic>.from(item as Map)))
          .toList(),
      usedFallback: json['used_fallback'] as bool? ?? false,
      emptyData: json['empty_data'] as bool? ?? false,
    );
  }
}

class FinanceSnapshotPayload {
  const FinanceSnapshotPayload({
    required this.currency,
    required this.transactions,
    required this.categories,
    required this.accounts,
    required this.budgets,
    required this.goals,
    required this.recurring,
  });

  final String currency;
  final List<Transaction> transactions;
  final List<Category> categories;
  final List<Account> accounts;
  final List<Budget> budgets;
  final List<Goal> goals;
  final List<RecurringTransaction> recurring;

  String get dataVersion {
    final stamps = [
      ...transactions.map((item) => item.updatedAt),
      ...budgets.map((item) => item.updatedAt),
      ...accounts.map((item) => item.updatedAt),
      ...goals.map((item) => item.updatedAt),
    ]..sort();
    return '${stamps.isEmpty ? 'empty' : stamps.last}:${transactions.length}';
  }

  Map<String, dynamic> toJson() => {
        'currency': currency,
        'transactions': transactions.map((item) => item.toJson()).toList(),
        'categories': categories.map((item) => item.toJson()).toList(),
        'accounts': accounts.map((item) => item.toJson()).toList(),
        'budgets': budgets.map((item) => item.toJson()).toList(),
        'goals': goals.map((item) => item.toJson()).toList(),
        'recurring': recurring.map((item) => item.toJson()).toList(),
      };
}

class InsightsApi {
  InsightsApi({http.Client? httpClient, String? baseUrl})
      : _http = httpClient ?? http.Client(),
        baseUrl = (baseUrl ?? AppEnv.current.insightsApiUrl).replaceAll(RegExp(r'/$'), '');

  final http.Client _http;
  final String baseUrl;

  Map<String, String> _headers(String accessToken, String userId) => {
        'Authorization': 'Bearer $accessToken',
        'Content-Type': 'application/json',
        'X-User-Id': userId,
      };

  Future<AnalyzeResponse> analyze({
    required String accessToken,
    required String userId,
    required String startDate,
    required String endDate,
    required FinanceSnapshotPayload snapshot,
  }) async {
    final response = await _http.post(
      Uri.parse('$baseUrl/insights/analyze'),
      headers: _headers(accessToken, userId),
      body: jsonEncode({
        'start_date': startDate,
        'end_date': endDate,
        'snapshot': snapshot.toJson(),
        'data_version': snapshot.dataVersion,
        'currency': snapshot.currency,
      }),
    );
    if (response.statusCode >= 400) throw Exception(await _error(response));
    return AnalyzeResponse.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }

  Future<ChatResponse> chat({
    required String accessToken,
    required String userId,
    required String message,
    required String startDate,
    required String endDate,
    required FinanceSnapshotPayload snapshot,
    String? conversationId,
    String? page,
  }) async {
    final response = await _http.post(
      Uri.parse('$baseUrl/insights/chat'),
      headers: _headers(accessToken, userId),
      body: jsonEncode({
        'message': message,
        'conversation_id': conversationId,
        'start_date': startDate,
        'end_date': endDate,
        'snapshot': snapshot.toJson(),
        'page': page,
      }),
    );
    if (response.statusCode >= 400) throw Exception(await _error(response));
    return ChatResponse.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }

  Future<String> _error(http.Response response) async {
    try {
      final body = jsonDecode(response.body) as Map<String, dynamic>;
      return body['detail']?.toString() ?? 'Insights request failed (${response.statusCode})';
    } catch (_) {
      return 'Insights request failed (${response.statusCode})';
    }
  }
}
